[inductor] Replace torch.allclose with torch.testing.assert_close in test_fx_fusion

robert-hardwick opened the pull request:
Preventative fix of a test failure with oneDNN v3.5 upgrade where order of float32 arithmetic may change in torch.admm ( bias term can be at the start or end of the arithmetic ) resulting in slightly different output due to float32 precision loss.

Replaced occurrences of torch.allclose with ~~torch._dynamo.testing.same~~  torch.testing.assert_close which is the recommended approach as per this issue https://github.com/pytorch/pytorch/issues/56544 ,the default tolerance is more relaxed than torch.allclose which satisfies the test with upcoming oneDNN change.

This should fix aarch64 ci failures in #129932 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @milpuz01

robert-hardwick committed:
[inductor] Replace torch.allclose with utility method in test_fx_fusion

Preventative fix of a test failure with oneDNN v3.5 upgrade where order of float32 arithmetic may change in torch.admm ( bias term can be at the start or end of the arithmetic ) resulting in slightly different output due to float32 precision loss.

Replaced occurrences of torch.allclose with torch._dynamo.testing.same
which is commonly used in other tests and has a lower default tolerance.

malfet commented:
@robert-hardwick please fix the lint?

malfet commented on test/inductor/test_fx_fusion.py:
```diff
@@ -66,7 +67,7 @@ def test_kwarg3(x, y):
         for f in [test_kwarg, test_arg, test_arg2, test_kwarg2, test_kwarg3]:
             traced = trace_func(f, inputs)
-            self.assertTrue(torch.allclose(f(*inputs), traced(*inputs)))
+            self.assertTrue(same(f(*inputs), traced(*inputs)))
```
Why not do something like that? And if lower tolerance is specific to arm platform, why not condition it on `platform.processor() == 'aarch64'`?
```suggestion
            self.assertEqual(f(*inputs), traced(*inputs), rtol=1e-4, atol=1e-4)
```

robert-hardwick replied:
I have fixed the lint and improved this by using `[torch.testing.assert_close](https://pytorch.org/docs/stable/testing.html#torch.testing.assert_close)` which seems to be the recommended approach. The default tolerances appears to satisfy all hardware types so no need to put a platform condition in.

robert-hardwick committed:
Use torch.testing.assert_close for tensor-based assertions in
test_fx_fusion

We use torch.testing.assert_close to compare tensors for likeness
which is the recommended approach as explained in this issue
 https://github.com/pytorch/pytorch/issues/56544

This uses a more relaxed default absolute tolerance of 1e-5 compared to torch.allclose which is necessary because test_permute_fusion may change the order of floating point operations.

yanbing-j commented:
Hi @malfet , could you please help review this PR? Thanks!

yanbing-j commented:
@pytorchbot rebase

malfet commented:
@pytorchbot merge

yanbing-j commented:
@pytorchbot  merge

malfet commented:
@pytorchbot merge
